### DISCUSSION TITLE
Create Organization is failing through REST API

### DISCUSSION THREAD

**Discussion by monojit18** [Sunbird-Ed/Community#186] [Installation] [answered]
Branch: release-4.10.1_RC1

I am not being able to Create an Organization using the REST API suite.
Although, I could create an user and can also retrieve token from Keycloak!

Following is the I/O from the CreateOrg call:

INPUT
====================================
curl --location --request POST 'http://sunbird.gcpwkshpdev.com/api/org/v1/create' \
--header 'X-Authenticated-User-Token: eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI4QU4tRVFKMkxNZVcwNG5wSVFrNnBMTlAzLTRLc3VqS0NtNGdlWXRBNFZVIn0.eyJqdGkiOiI4ZDA5NGIyOC0xZWZjLTQ4YTktYTgzNS02OTU0ZjEyMjY5YzciLCJleHAiOjE2NjM4MTE3MzgsIm5iZiI6MCwiaWF0IjoxNjYzNzY4NTM4LCJpc3MiOiJodHRwOi8vc3VuYmlyZC5nY3B3a3NocGRldi5jb20vYXV0aC9yZWFsbXMvc3VuYmlyZCIsImF1ZCI6WyJyZWFsbS1tYW5hZ2VtZW50IiwiYWNjb3VudCJdLCJzdWIiOiJhZDFiYTZiOS05ODY0LTQxZWYtOTcyZS02YWRkNTJiN2VkNjYiLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJsbXMiLCJhdXRoX3RpbWUiOjAsInNlc3Npb25fc3RhdGUiOiI3MWM2YTQxMy1iZjllLTRiNGUtYWNhYS1mZDBiMDA0NGQ3ZDYiLCJhY3IiOiIxIiwiYWxsb3dlZC1vcmlnaW5zIjpbImh0dHA6Ly9zdW5iaXJkLmdjcHdrc2hwZGV2LmNvbSJdLCJyZWFsbV9hY2Nlc3MiOnsicm9sZXMiOlsib2ZmbGluZV9hY2Nlc3MiLCJ1bWFfYXV0aG9yaXphdGlvbiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7InJlYWxtLW1hbmFnZW1lbnQiOnsicm9sZXMiOlsibWFuYWdlLWV2ZW50cyIsIm1hbmFnZS1yZWFsbSIsImNyZWF0ZS1jbGllbnQiLCJtYW5hZ2UtdXNlcnMiLCJtYW5hZ2UtYXV0aG9yaXphdGlvbiIsIm1hbmFnZS1jbGllbnRzIl19LCJsbXMiOnsicm9sZXMiOlsidW1hX3Byb3RlY3Rpb24iXX0sImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoiIiwiY2xpZW50SWQiOiJsbXMiLCJjbGllbnRIb3N0IjoiMjIzLjIyNS41OC4yMDAiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJzZXJ2aWNlLWFjY291bnQtbG1zIiwiY2xpZW50QWRkcmVzcyI6IjIyMy4yMjUuNTguMjAwIiwiZW1haWwiOiJzZXJ2aWNlLWFjY291bnQtbG1zQHBsYWNlaG9sZGVyLm9yZyJ9.WVHEFl7Hj17CdrF9J8kgkJ-ZImVifbOByM-14I1qPcHmFbRv-HtjgYGcMH5G4PhhOpdUih93srasTt1psG2nBJXF5FGo8fy5ztFB0QqPW_I7hMRCr_npCqICrLEsc6wfvdchznI1JYLh_Pvj8ZVP9Gl0sUgnN4HJoJw2S3NFaVXqFWAhxi_w6khxJaco7kdIScYx6LlquHSI4Bxk85FGldqMOMseeeEe0z6MnxO8LScVbJAQeC_XqrF0oQjaDJoqRCt2KtT5dd_LgrjukrjGKvST8ZwAqa7tZMyCAVOZVFvu6b2ybRgdoz-6LYQVJ4e6ll5kFK16BxJP-154Ssdvrw' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJHM1ZZMU5PWTc0ZktjRTMxdFVscFFVRnR0WGtETTFKeSJ9.RlCg2kysw-k5gTrEnYvPl0tzd7ow6nGJuyjF3GC_Cqo' \
--data-raw '{
    "request": {
        "orgName": "Sociology",
        "channel": "sunpersonal",
        "description": "First Channel",
        "isRootOrg": true,
        "isTenant": true,
        "organisationType": "board"
    }
}'

====================================

OUTPUT
====================================
{
    "id": null,
    "ver": "v1",
    "ts": "2022-09-21 13:57:34:850+0000",
    "params": {
        "resmsgid": null,
        "msgid": "1e79cfaa3cf7d07b9648da5707499b73",
        "err": "UNAUTHORIZED_USER",
        "status": "SERVER_ERROR",
        "errmsg": "You are not authorized."
    },
    "responseCode": "UNAUTHORIZED",
    "result": {}
}
====================================

As I have mentioned:

CreateUser and GetKeyCloakCredentials - both are working fine and returning respective responses.

Any help is much appreciated!

**Comment by beepdot** [comment, marked as answer]
Can you please check the logs from `learner` pods when you invoke this API and see what's happening? Maybe some logs will be able to provide us some hints.

**Reply by monojit18** [reply]
Hi @keshavprasadms,

My bad, I forgot to attach the learner pod log earlier, here it is when I try to CreateOrg:

[learner-log.txt](https://github.com/Sunbird-Ed/Community/files/9622785/learner-log.txt)

**Reply by beepdot** [reply]
Inside the learner pod, do you have a directory named `/keys`? Any files / contents inside this directory?

**Reply by monojit18** [reply]
@keshavprasadms 

I do not see any **keys** directory there!

This is what I see inside learner pod:

monojitdatta-macbookpro:sunbird-devops monojitdatta$ k exec -it learner-68b5c54647-9qkkd -n dev -- sh
/home/sunbird/learner $ ls
learning-service-1.0-SNAPSHOT           learning-service-1.0-SNAPSHOT-dist.zip

/home/sunbird/learner $ ls -la
total 126348
drwxr-sr-x    1 sunbird  sunbird       4096 Sep  7 14:30 .
drwxr-sr-x    1 sunbird  sunbird       4096 Sep 22 07:04 ..
drwxr-sr-x    1 sunbird  sunbird       4096 Sep 12 09:07 learning-service-1.0-SNAPSHOT
-rw-r--r--    1 root     root     129359510 Sep  7 14:30 learning-service-1.0-SNAPSHOT-dist.zip

/home/sunbird/learner $ cd learning-service-1.0-SNAPSHOT/
/home/sunbird/learner/learning-service-1.0-SNAPSHOT $ ls
RUNNING_PID  config       lib          start        start.bat

/home/sunbird/learner/learning-service-1.0-SNAPSHOT $

**Reply by monojit18** [reply]
@keshavprasadms:

Does this info help?
Am I missing anything from my end?

**Reply by beepdot** [reply]
Then you have missed some of the most important steps. Do all these variables have values? https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L52-L66. If not, please see the comments mentioned in the file and perform the steps. After which you have redeploy all these services - https://github.com/project-sunbird/sunbird-devops/blob/master/ansible/roles/stack-sunbird/defaults/main.yml#L1018-L1026 (Ignore the services which are not mentioned in the docc)

**Reply by beepdot** [reply]
Also `adminutils` needs to be redeployed.

**Reply by beepdot** [reply]
Also I asked to check for the directory `/keys`. Note that I have explicitly mentioned the path. You should check the same path starting from root directory. It seems you are checking in another directory. Please read instructions carefully.

`ls -larth /keys`

**Reply by monojit18** [reply]
Hi @keshavprasadms 

following is the output:

/home/sunbird/learner $ ls -larth /keys
total 4K     
lrwxrwxrwx    1 root     root          50 Sep 12 09:07 16lgJXcxImkIfU9EqCDBAXuDZUrYYbMZSiFBioqaMcU -> ..data/16lgJXcxImkIfU9EqCDBAXuDZUrYYbMZSiFBioqaMcU
lrwxrwxrwx    1 root     root          31 Sep 12 09:07 ..data -> ..2022_09_12_09_07_31.063629235
drwxr-xr-x    2 root     root          60 Sep 12 09:07 ..2022_09_12_09_07_31.063629235
drwxrwxrwt    3 root     root         100 Sep 12 09:07 .
drwxr-xr-x    1 root     root        4.0K Sep 12 09:07 ..

And the following file there, 
16lgJXcxImkIfU9EqCDBAXuDZUrYYbMZSiFBioqaMcU -  contains the core_vault_sunbird_sso_publickey value

**Reply by beepdot** [reply]
Can you reconfirm this step? 
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L63
Is the value of kid for RS256 token same as the value in variable? Can you please post a screenshot of the `Keys` page from keycloak admin page?

**Reply by monojit18** [reply]
Thanks @keshavprasadms 
Yes I have performed all these steps after KeyCloakRealm job.
But what I see now is that KId from Keycloak realm page is different than what I had entered. Not sure how that got changed!
If I now take the right Kid and update the Ansible variable in the repo, what other changes I need to do?
Can you please guide/suggest here?

If I have to re-run some of the pipelines, is there any risk of spoiling the current running env?

The screenshots for your reference

<img width="1550" alt="Screen Shot 2022-09-23 at 10 06 06 AM" src="https://user-images.githubusercontent.com/2395966/191892955-eebd79a1-b6cd-45ff-a270-e2281ffcb259.png">
<img width="1710" alt="Screen Shot 2022-09-23 at 10 06 19 AM" src="https://user-images.githubusercontent.com/2395966/191892967-22be638e-5118-43d8-bd76-c8d973c0cf4c.png">

**Reply by monojit18** [reply]
@keshavprasadms 
Can you please tell me if there are any pipelines that I have to re-run?
Is this list enough: https://github.com/project-sunbird/sunbird-devops/blob/master/ansible/roles/stack-sunbird/defaults/main.yml#L1018-L1026
And AdminUtils?

Is there any risk of spoiling the current running env?

Thanks in advance!

**Reply by beepdot** [reply]
Yes you can update the variable and redeploy these services mentioned. There will be no issues in doing so.

**Reply by beepdot** [reply]
Please also recheck other variables before doing so. SSO public key and refresh token variables.
https://github.com/project-sunbird/sunbird-devops/blob/master/private_repo/ansible/inventory/dev/Core/secrets.yml#L52-L66

**Reply by monojit18** [reply]
Thanks @keshavprasadms. This is resolved now!
Marking this as answered.
